scalableminds/webknossos
# Allow to add datasets through wk-connect

This PR adds an additional tab to the "Add Dataset" view, which allows to add a Neuroglancer dataset to wK using a webknossos-connect datastore. The Neuroglancer dataset is identified using the URL, all the layer information is parsed from that.

I'm not that happy with the code, as there's quite a bit of duplication between the Upload Dataset and Add wk-connect Dataset tabs, but I couldn't find a better abstraction to combine the two, maybe that's ok for now.

Issues that need to fixed:
- [x] wk-connect should include a max segment id for segmentation layers, otherwise the "Edit Dataset" view cannot be saved (@jstriebel)
- [x] There seems to be an issue with the datasource-properties.json of the default drosophila dataset. At least the frontend validation complains. Needs to be investigated.
  (both fixed in https://github.com/scalableminds/webknossos-connect/pull/64)
- [x] Proper testing of all dataset related functionality (does everything work with a wk-connect datastore)

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Set up an instance of webknossos-connect according to https://github.com/scalableminds/webknossos-connect#usage
- Open the "Add Dataset" view and switch to the "Add wk-connect Dataset" tab
  - Insert a dataset name, the datastore dropdown should be prefilled and paste the URL to a neuroglancer dataset (for example, the one I appended at the bottom)
    - You should be redirected to the import tab
  - Insert a bogus URL - You should not be able to click "Add" and get a suitable error message
- The Dataset Upload should still work as before
- The Dataset Upload during onboarding should also still work as before

### Issues:
- fixes #3744 

------
~~- [ ] Updated [documentation](../blob/master/docs) if applicable~~
~~- [ ] Needs datastore update after deployment~~


Example Neuroglancer dataset URL:
`https://neuroglancer-demo.appspot.com/#!%7B%22layers%22:%5B%7B%22source%22:%22precomputed://gs://neuroglancer-public-data/kasthuri2011/image%22%2C%22type%22:%22image%22%2C%22name%22:%22original-image%22%2C%22visible%22:false%7D%2C%7B%22source%22:%22precomputed://gs://neuroglancer-public-data/kasthuri2011/image_color_corrected%22%2C%22type%22:%22image%22%2C%22name%22:%22corrected-image%22%7D%2C%7B%22source%22:%22precomputed://gs://neuroglancer-public-data/kasthuri2011/ground_truth%22%2C%22type%22:%22segmentation%22%2C%22selectedAlpha%22:0.63%2C%22notSelectedAlpha%22:0.14%2C%22segments%22:%5B%2213%22%2C%2215%22%2C%222282%22%2C%223189%22%2C%223207%22%2C%223208%22%2C%223224%22%2C%223228%22%2C%223710%22%2C%223758%22%2C%224027%22%2C%22444%22%2C%224651%22%2C%224901%22%2C%224965%22%5D%2C%22name%22:%22ground_truth%22%7D%5D%2C%22navigation%22:%7B%22pose%22:%7B%22position%22:%7B%22voxelSize%22:%5B6%2C6%2C30%5D%2C%22voxelCoordinates%22:%5B5519.10400390625%2C8531.5380859375%2C1196.4942626953125%5D%7D%7D%2C%22zoomFactor%22:22.573112129999547%7D%2C%22perspectiveOrientation%22:%5B0.3849591314792633%2C0.6616792678833008%2C-0.6423912644386292%2C-0.0363389253616333%5D%2C%22perspectiveZoom%22:340.35867907175077%2C%22layout%22:%224panel%22%7D`
